pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,344 @@
+	if proxyDomain == "" || !strings.HasSuffix(hostname, proxyDomain) {
+		return out
+	}
+	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+".")
+	if proxyNamespace == nameinfo.Namespace {
+		out = append(out, nameinfo.Shortname+".")
+	}
+	// Do we need to generate entries for name.namespace.svc, name.namespace.svc.cluster, etc. ?
+	// If these are not that frequently used, then not doing so here will save some space and time
+	// as some people have very long proxy domains with multiple dots
+	// For now, we will generate just one more domain (which is usually the .svc piece).
+	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+"."+proxyDomainParts[0]+".")
+	return out
+}
+
+func (table *LookupTable) lookupHostIPv4(host string) []dns.RR {
+	ans := table.name4[host]
+	if len(ans) == 0 {
+		ans = table.cname[host]
+	}
+	return ans
+}
+
+func (table *LookupTable) lookupHostIPv6(host string) []dns.RR {
+	ans := table.name6[host]
+	if len(ans) == 0 {
+		ans = table.cname[host]
+	}
+	return ans
+}
+
+// This function stores the list of hostnames along with the precomputed DNS response for that hostname.
+// Most hostnames have a DNS response containing the A/AAAA records. In addition, this function stores a
+// variant of the host+ the first search domain in resolv.conf as the first query
+// is likely to be host.ns.svc.cluster.local (e.g., www.google.com.ns1.svc.cluster.local) due to
+// the list of search namespaces in resolv.conf (unless the app explicitly does www.google.com. which is unlikely).
+// We will resolve www.google.com.ns1.svc.cluster.local with a CNAME record pointing to www.google.com.
+// which will cause the client's resolver to automatically resolve www.google.com. , and short circuit the lengthy
+// search process down to just two DNS queries. This will eliminate unnecessary upstream DNS queries from the
+// agent, reduce load on DNS servers and improve overall latency. This idea was borrowed and adapted from
+// the autopath plugin in coredns. The implementation here is very different from auto path though.
+// Autopath does inline computation to see if the given query could potentially match something else
+// and then returns a CNAME record. In our case, we preemptively store these random dns names as a host
+// in the lookup table with a CNAME record as the DNS response. This technique eliminates the need
+// to do string parsing, memory allocations, etc. at query time at the cost of 2x number of entries (i.e. memory) to store
+// the lookup table.
+func (table *LookupTable) buildDNSAnswers(altHosts []string, ipv4 []net.IP, ipv6 []net.IP, searchNamespaces []string) {
+	for _, h := range altHosts {
+		if len(ipv4) > 0 {
+			table.name4[h] = a(h, ipv4)
+		}
+		if len(ipv6) > 0 {
+			table.name6[h] = aaaa(h, ipv6)
+		}
+		if len(searchNamespaces) > 0 {
+			// host h already ends with a .
+			// search namespace does not. So we append one in the end
+			randomHost := h + searchNamespaces[0] + "."
```

Comment:
do we only need `[0]` or all of them? Seems like just the first, but wasn't sure if this is how autopath worked or if there is any reason th first is not sufficient

Reply:
So coredns maintains a trie like structure for all names and then traverses through it. Essentially its a memory tradeoff. In our case, its a flat hash map. And we catch the thing on the very first search query. If the client does parallel dns resolution over all search namespaces, this may not work. In that scenario, we would have to add an entry for each search namespace. I fear that it would increase the memory overhead though. For 5 entries in search namespaces, the map of 1000 entries will become 5000..

Reply:
Makes sense. Worst case this is an optimization anyhow